README
Hydna for nodejs
================

The node.js hydna library is a straight-forward implementation of the wink binary protocol. The module has been closely modeled to the native node.js modules and the usage pattern should come naturally to developers using other node.js modules. The module does not have any external dependencies.

Creating a connection:

    var hydna = require('hydna');
    var channel = hydna.createChannel('localhost', 'rw');

    channel.on('connect', function() {
      // read/write connection is ready to use 
    });

    channel.on('error', function() {
      // an error occured when connecting
    });

A read/write channel is opened and event-listeners for connect and error are attached to the stream.

Sending Data:

    var hydna = require('hydna');
    var channel = hydna.createChannel('localhost', 'w');

    channel.on('connect', function() {
      var message = 'Hello World!';
      channel.write(message, 'utf8');
    });

Opens up a channel for writing and, when the connection has been established and the connect event has been emitted, writes a message.

Receiving data:

    var hydna = require('hydna');
    var channel = hydna.createChannel('localhost', 'r');
    
    stream.setEncoding('utf8');
    
    stream.on('data', function(data) {
      console.log(data);
    });

Opens up a channel for reading and, when the data arrives and the data event is emitted, writes the received data to the console.


## Installation

The easiest way to install this library is via npm:

    $ npm install hydna


## Test 

The test suite requires a running server instance with the behaviors found `test/behaviors` loaded.

The default server address is `localhost:7010`. This can be overridden via the environment variable `TEST_ADDRESS`.

To run the test suite:

    $  TEST_ADDRESS=<server_address_and_port> npm test 


## API Specification

The node.js hydna client library consists of a single module — hydna — that
contains everything that is necessary to communicate over hydna.

### hydna.PAYLOAD_MAX_SIZE

Returns the maximum payload size.

### hydna.followRedirects=true

Indicates if HTTP-redirects should be followed or not.


### hydna.origin=<hostname>
  
Sets the origin identity that should be sent to the server on handshake.


### hydna.agent=node-winsock-client/<version>
  
Sets the agent identity that should be sent to the server on handshake.


### hydna.createChannel(url, mode, [callback])

Opens a Channel to the specified ´'url'´.

This function is asynchronous. When the `'connect'` event is emitted once connected. If there is a problem connecting, the `'connect'` event will not be emitted, the 'error' event will be emitted with the exception.

Available modes:
* read (r) - Open channel in read mode
* write (w) - Open channel in write mode
* readwrite (rw) - Open channel in read-write mode.
* +emit(e) - Open channel with emit-signal support (e.g. 'rw+emit').

This example opens a Channel and writes data too it. The same data
is received :

    var hydna = require('hydna');
    var channel = hydna.createChannel('localhost', 'w');
    channel.write('Hello World!');


### hydna.Channel

This object is an abstraction of of a TCP or UNIX socket. hydna.Channel  instance implement a duplex stream interface. They can be created by the user and used as a client (with connect()) or they can be created by Node and passed to the user through the 'connection' event of a server.

hydna.Channel instances are EventEmitters with the following events:

Event: `'connect'`
`function (message) { }`

Emitted when a channel connection successfully is established. See connect(). The `'message'` argument may or may not contain an initial message from server.

Event: `'data'`
`function (data) { }`

Emitted when data is received. The argument data will be a Buffer or String.  Encoding of data is set by channel.setEncoding(). 

Event: `'drain'`
`function () { }`

Emitted when the write buffer becomes empty. Can be used to throttle uploads.

Event: `'error'`
`function (exception) { }`

Emitted when an error occurs. The `'close'` event will be called directly  following this event.

Event: `'close'`
`function (had_error) { }`

Emitted once the channel is fully closed. The argument had_error is a boolean which says if the channel was closed due to an error.

Event: `'signal'`
`function (data) { }`

Emitted when remote server send's a signal.

#### Channel.readable

Returns `true` if channel is readable, else false.

#### Channel.writable

Returns `true` if channel is writable, else false.

#### Channel.emitable

Returns `true` if channel is emitable, else false.


#### Channel.readyState

Either `'closed'`, `'closing'`, `'open'`, `'opening'`, 
`'read'`, `'write'`, `'readwrite'` and/or `'+emit'`.


#### Channel.id

Returns channel `id` as a string. Property is `null` if not connected.


#### Channel.setEncoding(encoding=null)

Sets the encoding (either `'ascii'`, `'utf8'`, `'base64'`, `'json'`)


#### Channel.connect(url, mode='readwrite')

See `hydna.createChannel` for more info.


#### Channel.write(data, encoding='ascii', priority=1)

Sends data on the channel. The second parameter specifies the encoding in the case of a string--it defaults to ASCII because encoding to UTF8 is rather slow.

Returns ´true´ if the entire data was flushed successfully to the underlying connection. Returns `false` if all or part of the data was queued in user memory. ´'drain'´ will be emitted when the buffer is again free.


#### Channel.dispatch(message)

Dispatch a signal on the channel.

Returns ´true´ if the signal was flushed successfully to the underlying connection. Returns `false` if the all or part of the signal was queued in user memory. ´'drain'´ will be emitted when the buffer is again free.

Example:

    var hydna = require('hydna');
    var channel = hydna.createChannel('localhost', 'read+emit');
    channel.on('signal', function(message) {
      if (message == 'pong') {
        console.log('Recevied pong from server');
      }
    });
    channel.dispatch('ping');



#### Channel.end([message])

Closes channel for reading, writing and emitting. The optional `message` is sent to the server.

Example:

    var hydna = require('hydna');
    var channel = hydna.createChannel('localhost', 'read');
    channel.end('good bye!');


#### Channel.destroy()

Closes channel for reading, writing and emitting.

CHANGELOG
## v2.0.3 (2015-04-07)

Changes:

  - Made library compatible with Node v0.12.X


## v2.0.1 (2014-02-06)

Changes:

  - Activate HEARBEAT at connect


## v2.0.1 (2014-02-06)

Bugfixes:

  - Fixed bug when error was catched and never display for user
  - Fixed bug where channel tried to connect to socket multiple times
  - Fixed a refcount bug that could happen if createChannel throwed an
    exception

## v2.0.0 (2013-11-20)

Features:

  - Fully compatible with wink protocol 1


## v1.2.3 (2013-11-19)

Features:

  - Added option to disable multiplexing when creating a channel (useful when
    testing many connection from one process).

Bugfixes:

 - Fixed issue where channels could be resolved multiple times
 - Fixed issue with ghost http requests
 - Minor bug fixes


## v1.2.2 (2013-10-30)

Features:

  - Added Multi-connection test
  - Added Close-data test

Bugfixes:

  - Fixed issue where signals was not transmitted if channel did not open in
    emit mode.


## v1.2.1 (2013-10-29)

Features:

  - Full TLS support

Bugfixes:

  - Fixed issue with https in push API.


## v1.2.0 (2013-10-27)

Features:

  - Core completly rewritten to be more stabel and logical.
  - Exposes two new Error classes, OpenError and SignalError. Useful to
    determind if channel was closed by an OpenDeny.


## v1.1.1 (2013-10-21)

Features:

  - Removed obsolete code.
  - Added examples: helloworld, listener, mulitplechannels and signals

Bugfixes:

  - Fixed broken test-suite.
  - Fixed issue with broken token
  - Fixed issue with path not being registered


## v1.1.0 (2013-10-11)

Features:

  - Support for new path-based protocol
  - Added support for UTF8 and Binary frames.
  - Removed obsolete code


## v1.0.9 (2013-09-16)

Features:

  - Added interface for "push"


## v1.0.8 (2013-09-09)

Features:

  - Added keep-alive timer


test/A.js
var sys = require("sys");
var wink = require("../lib/wink");


var stream = wink.open("AABBCCDDEEFF11", "rw");


stream.addListener("connect", function() {
  stream.write("Hello world!");
});


stream.addListener("data", function(data) {
  sys.puts("Skaggi säger: " + data);
});

test/basic.js
var sys = require("sys");
var wink = require("../lib/wink");

var stream = wink.open("aabbccdd11112222", "rw");

stream.addListener("connect", function() {
  sys.puts("connected to server, now sending data");
  stream.write("hejsan svejsan", "ascii");
});

stream.addListener("data", function(data) {
  sys.puts("recived data: " + data.toString());
});


test/multi.js
const spawn = require("parall/worker").spawn;


spawn("./receiver.js", 10, [], function() {
  console.log("start");
  spawn("./sender.js", 10);
});

behaviors/signal.js
//
//        Copyright 2011 Hydna AB. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//    1. Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//    2. Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//  THIS SOFTWARE IS PROVIDED BY HYDNA AB ``AS IS'' AND ANY EXPRESS OR IMPLIED
//  WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
//  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
//  EVENT SHALL HYDNA AB OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
//  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
//  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
//  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
//  ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
//  TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//  The views and conclusions contained in the software and documentation are
//  those of the authors and should not be interpreted as representing
//  official policies, either expressed or implied, of Hydna AB.
//

var request = require("request");
var signal = require("signal");

if (request.getToken() == "ping") {
  signal.reply("pong");
} else {
  signal.reply("bad token");
}

examples/signals.js
var hydna = require('../index');

// Open channel in "emit" mode. We are requesting to "emit" signal to the
// underlying behavior-layer.
var channel = hydna.createChannel('public.hydna.net/ping-back', 'emit');

// Connect handler, is triggered once that the channel is connected
channel.on('connect', function (data) {
  
  // Dispatch a signal with message "ping". The underlying behavior-layer
  // will recieve it and can choose to take action
  this.dispatch('ping');
});

// Signal handler, is triggered each time a SIGNAL is received on the channel
channel.on('signal', function (data) {

  // We received a signal on our disptach.
  if (typeof data == "string") {
      console.log('Received singal "%s"', data);
  } else {
      console.error("Expected a pong response");
  }

  // Close the channel, which terminates the underlying
  // receive-loop.
  this.end();
});

examples/listener.js
var hydna = require('../index');

// Open channel in "read" mode. We are just listening for data and signals.
var channel = hydna.createChannel('public.hydna.net', 'read');

// Connect handler, is triggered once that the channel is connected
channel.on('connect', function (data) {
  console.log("Press Ctrl-C to abort the receive loop");

  // The "connect" handler may contain an optional welcome-message sent
  // by remote part. Print message if exists.
  if (typeof data == 'string') {
    console.log("[WELCOME]: " + data);
  }
});

// Data handler, is triggered each time DATA is received on the channel
channel.on('data', function (data) {
  if (typeof data == 'string') {
    console.log('[DATA] ' + data);
  } else {
    console.log('[DATA] <binary>');
  }
});

// Signal handler, is triggered each time a SIGNAL is received on the channel
channel.on('signal', function (data) {
  if (typeof data == 'string') {
    console.log('[SIGNAL] ' + data);
  } else {
    console.log('[SIGNAL] <binary>');
  }
});

examples/multiplechannels.js
var hydna = require('../index');

var messages = ['Hello', 'World'];

// Open two channels in "readwrite" mode.
var channels = [
  hydna.createChannel('public.hydna.net/channel1', 'rw'),
  hydna.createChannel('public.hydna.net/channel2', 'rw')
];

// Loop through both channels and add handlers.
channels.forEach(function (channel) {

  // Connect handler, is triggered on both channels, once connected
  channel.on('connect', function () {
    console.log('Connected to channel "%s"', this.url);

    // Write messages in correct order.
    this.write(messages.shift());
  });

  // Data handler, is triggered on both channels, once DATA is received
  channel.on('data', function (data) {
    console.log('Channel "%s" recieved: %s', this.url, data);

    // Close the channel. The application will terminate once both
    // channels are closed.
    this.end();
  });

});

lib/wink.js
//
// Interface for working with Wink Remote TCP connections
//
var sys     = require("sys"),
    net     = require("net");
    Buffer  = require("buffer").Buffer;

var EventEmitter  = require("events").EventEmitter;
    
const PAD_NONE = exports.PAD_NONE = 'none';
const PAD_LEFT = exports.PAD_LEFT = 'left';
const PAD_RIGHT = exports.PAD_RIGHT = 'right';

const DEFAULT_PORT = 7010;
const DEFAULT_HOST = "127.0.0.1";

const HEADER_LENGTH = 12;    

// OTHER
const PING      = 0x01;  // Ping

// CLIENT COMMANDS
const OPENL     = 0x02;  // Open Live
const OPENBL    = 0x03;  // Open BackLog
const EMIT      = 0x04;  // Emit to a channel
const CLOSE     = 0x05;  // Close a Channel without closing the actuall socket.

// SERVER EVENTS
const OPENSTAT  = 0x06;  // Open STATUS
const RTDATA    = 0x07;  // Realtime data
const DATA      = 0x08;  // Time-based data
const INTERRUPT = 0x09;  // A channel is interrupted

// Channel modes
const READ      = 0x01;
const WRITE     = 0x02;
const READWRITE = 0x03;

const BIN = "binary";

const NULL_ADDR = "\x00\x00\x00\x00\x00\x00\x00\x00";

var connections = {};

/**
 *  Open a stream
 */
exports.open = function(hexaddr, mode, options) {
  options = options || {};
  mode = mode || "r";
  var addr = exports.binaddr(hexaddr);
  var port = options.port || DEFAULT_PORT;
  var host = options.host || DEFAULT_HOST;
  var live = options.live || true;
  var token = options.token || null;
  var serverid = host + ':' + port;
  var connection = connections[serverid];
  var stream = null;
  
  function sendOpen() {

    if (live) {
      var openPacket = new Buffer(13 + (token ? token.length : 0));
      openPacket[0] = Math.floor(openPacket.length / 256) & 0xff;
      openPacket[1] = openPacket.length % 256;
      openPacket[2] = OPENL;
      openPacket[3] = 0;
      openPacket.write(addr, "binary", 4)
      openPacket[12] = mode == "r" ? READ : mode == "w" ? WRITE : READWRITE;

      if (token) {
        openPacket.write(token, "ascii", 13);
      }

      connection.write(openPacket);
    }
  }
  
  if (!connection) {
    connection = createServerConnection(port, host, function() {
      sendOpen();
    });
    
    connections[serverid] = connection;
  } else {
    if (connection.readyState == "opening") {
      connection.addListener("connect", function() {
        sendOpen();
      });
    } else {
      sendOpen();
    }
  }
  
  if (connection.openStreams[addr]) {
    throw new Error("Stream is already opened.");
  }
  
  stream = new Stream(connection, addr, live, mode);
  
  stream.addListener("close", function() {
    connection.openStreams[addr] = undefined;
  })

  connection.openStreams[addr] = stream;
  
  return stream;
}

/**
 *  Converts a binaddr into a hexaddr
 */
exports.hexaddr = function(addr) {
  if (addr == null) {
    return hexaddr(ADDR_NULL);
  }
  var index = addr.length;
  var result = [];
  while (index--) {
    var comp = (addr.charCodeAt(index) + 
                addr.charCodeAt(--index) * 256).toString(16);

    while (comp.length < 4) {
      comp = '0' + comp;
    }
    
    result.unshift(comp);
  }
  return result.join(':');
}

/**
 *  Converts a hexaddr into a binaddr
 */
exports.binaddr = function(hex_addr, pad) {
  var hex = hex_addr.replace(/\:/g, '');
  var padding = pad || PAD_NONE;
  
  if ([PAD_LEFT, PAD_RIGHT, PAD_NONE].indexOf(padding) == -1) {
    throw 'Uknown padding method: ' + padding;
  }
  
  if (hex.length > 16) {
    throw 'Key is larger then 16 chars';
  }
  
  if (padding != PAD_NONE) {
    while (hex.length < 16) {
      hex = padding == PAD_LEFT ? '0' + hex : hex + '0';
    }
  }
  
  var buffer = [];
  var comp = null;
  
  while (hex.length && (comp = hex.substr(0, 4))) {
    var intComp = parseInt(comp, 16);
    buffer.push(String.fromCharCode(Math.floor(intComp / 256) & 0xff));
    buffer.push(String.fromCharCode(intComp % 256));
    hex = hex.substr(4);
  }
  
  return buffer.join('');
}

function Stream(conn, addr, live, mode) {
  var self = this;
  
  self.readyState = "opening";
  self.addr = addr;
  self.originalAddr = addr;
  self.readable = (mode.indexOf("r") != -1);
  self.writable = (mode.indexOf("w") != -1);
  self.live = live;
  
  self.addListener("connect", function() {
    self.readyState = self.readable && self.writable ? "open" : 
                          self.readable ? "readOnly" : "writeOnly";
    
  });

  self.addListener("end", function(code) {
    self.readyState = "closed";
    self.readable = false;
    self.writable = false;
  });
  
  self.conn = conn;
}

sys.inherits(Stream, EventEmitter);

/**
 *  Write a message to stream
 */
Stream.prototype.write = function(data, encoding) {

  if (!this.writable) {
    throw new Error("Stream is not writable");
  }
  
  var header = new Buffer(12);
  var length = typeof data === "string" ? Buffer.byteLength(data, encoding) : data.length;
  length += header.length;
  header[0] = Math.floor(length / 256) & 0xff;
  header[1] = length % 256;
  header[2] = EMIT;
  header[3] = 0;

  header.write(this.addr, "binary", 4);
  
  this.conn.write(header);
  
  if (typeof data === "string") {
    this.conn.write(data, encoding || "binary");
  } else {
    this.conn.write(data);
  }  
}

Stream.prototype.end = function(data) {
  if (this.readyState == "closed") {
    return;
  }
  
  if (data) {
    this.write(data);
  }

  var header = new Buffer(12);
  header[pos] = Math.floor(header.length / 256) & 0xff;
  header[pos + 1] = header.length % 256;
  header[pos + 2] = CLOSE;
  header[pos + 3] = 0;

  header.write(this.addr, "binary", pos + 4);
  self.conn.write(data);
  self.emit("close", 0);
}

function createServerConnection(port, host, callback) {
  var connection = net.createConnection(port, host);
  var READ_BUFFER_LENGTH = 4096 * 3;
  var buffer = new Buffer(READ_BUFFER_LENGTH);
  var bufferLen = 0;
  var bufferPos = 0;
  var headerParsed = false;
  var packetLength = 0;
  var packetFlag = 0;
  var packetAddr = null;
  var header = null;
  
  // Used for stats
  connection.bufferAllocs = 0;
  connection.bufferCopies = 0;
  connection.bufferMaxSize = 0;

  connection.setTimeout(0);
  connection.setNoDelay(true);
  
  connection.openStreams = {};
  
  connection.addListener("connect", callback);

  connection.addListener("data", function(data) {
    var tempBuffer = null;
    var currentSize = null;
    
    if (bufferLen + data.length > buffer.length) {
      currentSize = bufferLen - bufferPos;
      tempBuffer = new Buffer(Math.max(currentSize + data.length, READ_BUFFER_LENGTH));
      if (bufferPos != bufferLen) {
        buffer.copy(tempBuffer, 0, bufferPos, bufferLen);
        bufferLen = bufferLen - bufferPos;
      } else {
        bufferLen = 0;
      }
      
      bufferPos = 0;
      buffer = tempBuffer;
    }
    
    data.copy(buffer, bufferLen, 0, data.length);
    bufferLen += data.length;
    
    while (bufferPos < bufferLen) {

      if (!headerParsed) {

        if (bufferPos + HEADER_LENGTH > bufferLen) {
          return;
        }
        
        packetLength = (buffer[bufferPos] * 256) + 
                            buffer[bufferPos + 1];
        packetFlag = buffer[bufferPos + 2];
        packetReserved = buffer[bufferPos + 3];
      }

      if (bufferPos + packetLength > bufferLen) {
        return;
      }
      
      var packetAddr = buffer.toString(BIN, bufferPos + 4, bufferPos + 12);

      switch (packetFlag) {
        
        case OPENSTAT:
        
          if (connection.openStreams[packetAddr]) {
            var stream = connection.openStreams[packetAddr];
            
            var code = buffer[bufferPos + 12];
            var responseAddr = buffer.toString(BIN, bufferPos + 13, bufferPos + 21);

            switch (code) {

              case 1:
                stream.emit("connect", code, packetAddr);
                break;
                
              case 2:
                stream.addr = responseAddr;
                stream.emit("connect", code, responseAddr);
                break;
              
              default:
                stream.emit("error", code);
                break;
            }
          }
          break;
          
        case INTERRUPT:
          // TODO:
          break;
          
        case RTDATA:
          if (connection.openStreams[packetAddr]) {
            var packetSlice = buffer.slice(bufferPos + 12, bufferPos + packetLength);
            connection.openStreams[packetAddr].emit("data", packetSlice);
          }
          break;
      }

      bufferPos += packetLength;
      header = null;
    }
  });
  
  connection.addListener("end", function() {
  });
 
  return connection;
}